### ISSUE TITLE
Investigate effect of `async` attribute on `type="module"` inline scripts

### ISSUE THREAD

**Issue by romaricpascal** [alphagov/govuk-frontend#3367]
## What

Test using `async` on a `type="module"` `<script>` tag for setting a class highlighting that ES modules are supported.
Check if:
- it does run immediately or so, at least ahead of "regular" `type="module"` scripts that are deferred by default
- it runs soon enough in regards to the DOM construction that there's no visible changes of styles

## Why

We're looking to use `<script type="module">` as a way to have scripts stop loading in older browsers. We need a CSS class on the document to detect if ES modules are available (and as such, our components will be initialised). Currently, we're [setting a `js-enabled` class with an inline script in the document's `<head>`](https://frontend.design-system.service.gov.uk/get-started/#5-get-the-javascript-working), but we need an extra class that highlights the availablility of ES modules specifically.

That class needs to be set as  soon as possible, to avoid a flash of the page styled as if JS was not available. An inline `<script type="module">` would run only after the whole DOM is built, so would cause such a flash, so we need an alternative solution.

## Timebox

We should review progress after this period of time has elapsed, even [if the spike has not been 'completed'](https://docs.google.com/document/d/17W_d0aYszgh7HTgnTHZLxXDu_TJyu8zzWt_lXH8VaHA/edit#heading=h.jd56itytegn6)

## Who is working on this?

Spike lead: @colinrotherham

Spike buddy: @romaricpascal

## Questions to answer

- [ ] Does the `async` attribute let us execute a module script straitght away?
	- [x] If not, which other routes are available? (regular `<script>` with `'nomodule' in HTMLScriptElement.prototype`?, though it may be `false` in browsers that do support `<script type="module">`, like Safari 10.1)

## Done when

You may find it helpful to refer to our [expected outcomes of spikes](https://docs.google.com/document/d/17W_d0aYszgh7HTgnTHZLxXDu_TJyu8zzWt_lXH8VaHA/edit#heading=h.mmkqzigd11rs).

- [x] Questions have been answered or we have a clearer idea of how to get to our goal
- [x] Findings have been reviewed and agreed with at least one other person
- [x] Findings have been shared, e.g: via a write-up on the ticket, at a show & tell or team meeting

**Comment by colinrotherham** [comment]
Added to dev catchup to review findings

**Comment by colinrotherham** [comment]
Reviewed at dev catchup this morning

Due to `<script type="module">` deferred execution we've agreed `<script>` is our only option

Even module execution in `<head>` targeting `<html class="govuk-template js-enabled">` didn't prevent the flicker

### Assumptions

1. **Classes will always be present on `<body>`**
  Unlike [`govuk_template`](https://github.com/alphagov/govuk_template/blob/61f78b535872c2a46ceec7f2793236a4ae96771c/source/views/layouts/govuk_template.html.erb#L39) (deprecated) where body classes were optional

2. **Detection script should not crash in older browsers**
  Which means we can't use `.classList.add()` with multiple classes

## Before
Previously, the class `.js-enabled` was either appended or set as the `<body>` class

```html
<script>
  document.body.className = (
    document.body.className
      ? document.body.className + ' js-enabled'
      : 'js-enabled'
  );
</script>
```

## After
Here's a potential candidate

```html
<script>
  document.body.className += ' js-enabled' + (
    'noModule' in HTMLScriptElement.prototype
      ? ' mjs-enabled'
      : ''
  )
</script>
```
